src/file/create.h

```diff
@@ -0,0 +1,169 @@
+    push_file_warning(ctx, &fm, FM_CREATE);
+
+    NoEvent:
+    return;
+}
+
+static __always_inline void exit_create(struct pt_regs *ctx)
+{
+    u32 key = 0;
+    buf_t *buffer = (buf_t *)bpf_map_lookup_elem(&buffers, &key);
+    if (buffer == NULL) return;
+
+    cached_path_t *cached_path = (cached_path_t *)bpf_map_lookup_elem(&percpu_path, &key);
+    if (cached_path == NULL) return;
+
+    file_message_t *fm = (file_message_t *)buffer;
+    cursor_t cursor = { .buffer = buffer, .offset = &fm->u.action.buffer_len };
+    int ret = 0;
+    error_info_t info = {0};
+
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+    incomplete_create_t *event = bpf_map_lookup_elem(&incomplete_creates, &pid_tgid);
+    if (event == NULL) return;
+    if (event->pid_tgid != pid_tgid) goto EventMismatch;
+
+    if (cached_path->next_dentry != NULL) {
+        if (event->target_dentry == NULL) goto ResolveTarget;
+        else goto ResolveSource;
+    }
+
+    if (event->target_dentry == NULL || event->target_vfsmount == NULL) {
+        set_empty_local_warning(W_NO_DENTRY);
+        goto EmitWarning;
+    }
+    ret = extract_file_info_owner(event->target_dentry, &fm->u.action.target, &fm->u.action.target_owner);
+    if (ret < 0) goto EmitWarning;
+
+    fm->type = FM_CREATE;
+    fm->u.action.pid = event->pid_tgid >> 32;
+    fm->u.action.mono_ns = event->start_ktime_ns;
+    fm->u.action.buffer_len = sizeof(file_message_t);
+    fm->u.action.u.create.source_link = event->link_type;
+
+    init_filtered_cached_path(cached_path, event->target_dentry, event->target_vfsmount);
+
+    ResolveTarget:
+    ret = write_path(ctx, cached_path, &cursor, EXIT_CREATE);
+    if (ret < 0) goto EmitWarning;
+    write_null_char(cursor.buffer, cursor.offset);
+    fm->u.action.tag = (cached_path->filter_state >= 0) ? cached_path->filter_tag : -1;
+    switch (event->link_type) {
+        case LINK_NONE:
+            if (cached_path->filter_state <= 0) goto NoEvent; // Didn't match watched path filter
+            else goto FinishMessage;
+        case LINK_SYMBOLIC:
+            if (cached_path->filter_state <= 0) goto NoEvent; // Didn't match watched path filter
+            write_string(event->source, cursor.buffer, cursor.offset, PATH_MAX);
+            write_null_char(cursor.buffer, cursor.offset);
+            goto FinishMessage;
+        case LINK_HARD:
+            event->target_dentry = NULL; // mark that we are done with target
+            init_filtered_cached_path(cached_path, event->source, event->target_vfsmount);
+            // don't do the filter logic again if we already matched on target
+            if (fm->u.action.tag != -1) cached_path->filter_state = -1;
+    }
+
+    ResolveSource:
+    write_path(ctx, cached_path, &cursor, EXIT_CREATE);
+    if (fm->u.action.tag == -1) { 
+        if (cached_path->filter_state <= 0) goto NoEvent;
+        fm->u.action.tag = cached_path->filter_tag;
+    }
+    write_null_char(cursor.buffer, cursor.offset);
```

Comment:
I don't think this is necessary since we are done writing paths (although I don't think it hurts)